lsコマンドを作る１(オプションなしのlsコマンド)のls.rbプログラムを作成

a-kuroki-gs opened the pull request.

a-kuroki-gs committed:
lsコマンドを作る１のls.rbをコミットします

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,111 @@
+
+OptionParser.new
+
+# 標準入力を受け取って変数に代入
```
慣れてきたらでいいですが、コメントは基本的に必要な箇所でのみ書くようにするといいかなと思います！

### 書かない
- コードと同じことを日本語で言い換えているだけのコメント
  ```ruby
  1 + 2 # 1と2を足す
  ```
- コード中に書いてもしょうがないこと
  ```ruby
  # いつかリファクタリングする
  ```
  これきっかけでやることはほぼないので、タスク管理ツールにでも書きましょう。
  せめて `TODO: `とつけておくと、ツールなどでTODOだけ抽出したりすることはできます。

### 書くとよい
基本的にコードから読み取れないことを書きます。

- 目的や背景などの説明
  ```ruby
  # ○○のために必要
  def xxx
  ```
  これもメソッド名などをそのままの言い換えているだけなら不要

- インターフェースの説明
  ```ruby
  # 引数aは { name: String, age: Integer } という形を想定している
  def xxx(a)
  ```
  - こちらは[rdoc](https://github.com/ruby/rdoc)や[yard](https://yardoc.org/)などの形式で書くと汎用性が高いです。
  - [rbs](https://github.com/ruby/rbs)の型情報として書くという手もあります

- 設計判断のコンテキスト
  ```ruby
  # 通常○○を使うが、ここでは△△が重要になってくるため、こういう設計にした
  ```

a-kuroki-gs replied:
不要なコメントは削除するように以後気を付けます。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,111 @@
+      normal_dir << target
+    end
+  end
+  colon_dir_all << colon_dir
```
これって使われていないですかね？
不要であれば消してしまったほうがよいかと思います。

a-kuroki-gs replied:
こちら`-a`オプションをつける際に利用する予定で残しておりました。
`#TODO:`として、コメントでその旨記しておきます。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,111 @@
+normal_dir_all = normal_dir_all.map(&:sort)
+
+# 表示する列数に応じて行数を計算するメソッド
+def calc_num_of_lines(element_count, col)
+  # ファイル数をcolで割って余りが0ならelement_count行
+  if (element_count % col).zero?
+    element_count / col
+  # ファイル数をcolで割って余りが0でなければelement_count + 1 行
+  else
+    element_count / col + 1
+  end
+end
```
ここは Ruby3.2以降から導入された `ceildiv` を使うと簡単に書けるかなと思います。
https://docs.ruby-lang.org/ja/latest/method/Integer/i/ceildiv.html

a-kuroki-gs replied:
`ceildiv`知らなかった＆見つけられなかったので勉強になりました！

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,111 @@
+
+# 要素を表示するメソッド
+def print_elements(element)
+  # ３列表示の時の行数を計算
+  max_row_length = calc_num_of_lines(element.size, 3)
+
+  modified_elements = []
+  max_name_size_array = []
+
+  # 行として分割
+  element.each_slice(max_row_length) do |e|
+    # 行(のちの列)ごとに最大ファイル文字数をカウント
+    max_name_size_array << e.map(&:length).max
+    # 要素が足りないところにはnilを代入
+    modified_element = e + [nil] * (max_row_length - e.size)
+    modified_elements << modified_element
+  end
+  element = modified_elements
+
+  # 要素を転置
+  element = element.transpose
+
+  # 最大ファイル文字数+2の幅で、nil以外の配列の要素を表示
+  print_element(element, max_name_size_array)
```
メソッド分割は抽象度を揃えるとよいです。
今だと `print_element` だけ抽象的なメソッドがあり、他の部分は内部の実装まで見えてしまっています。

たとえば、「カレーを作る」という工程をRubyのメソッドで表現したいとして、まず日本語で抽象度をそろえて分割してみます。
```
- カレーを作る
  - 材料の下ごしらえ
    - 玉ねぎを切る
    - にんじんを切る
    - じゃがいもを切る
  - 材料に火を通す
    - 肉、野菜を炒める
    - 水を入れて煮込む
  - 味つけをする
    - ルーをいれる
```

これをもとに、

```ruby
def make_curry
  prepare_ingredients
  cook_ingredients
  season
end
```

のようにメソッドを作るとよいです。
たとえば、

```ruby
def make_curry
  chop_onions
  chop_carrots
  chop_potatoes
  cook_ingredients
  season
end
```

のようになっていると材料を切るところだけ抽象度が低く、ちぐはぐなメソッドになってしまいます。

a-kuroki-gs replied:
ここのメソッドの抽象度を揃えるところが難しかったです。
`print_elements`メソッドの中身を
①表示する配列を作成する(行数カウント→行ごとに分割→配列を転置)
②①で作成した配列を行ごとに表示
という流れにしましたが、そのような考え方でよいでしょうか？

yoshitsugu replied:
わかりやすくなったと思います！

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,111 @@
+
+# lsコマンドと同じように表示
+normal_dir_all.each_with_index do |normal, idx|
+  if normal_dir_all.size > 1 || file_on_input == true
```
`== true` は `true` か `false` を返すので、余程 `true` とそれ以外に `true` と同じように `if` の条件としては振る舞う値を区別したい場合以外は省略して大丈夫です。

a-kuroki-gs replied:
その通りですが抜けておりました。。。
`== true`が他のところにもありましたので、どちらも削除しました！

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,111 @@
+  print_element(element, max_name_size_array)
+end
+
+# lsコマンドと同じように表示
```
ここのコメントはあまり意味ないかなと思います。
全体が `ls` コマンドと同じように表示しましょう、というプラクティスなので。

a-kuroki-gs replied:
削除しました。

yoshitsugu reviewed (commented):
歓迎要件にもとりくまれていてよいですね :+1:

a-kuroki-gs committed:
不要なコメントや記述の削除＆メソッドの抽象度を合わせて変更しました

不要なコメントや==trueを削除
自作のcalc_num_of_linesメソッドを削除し、ceildivメソッドを利用して簡単に表記
print_elementsの中身の抽象度を合わせて、メソッドの削除・追加・変更

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,89 @@
+
+input_all << '.' if input_all == []
+
+target_dir_all = []
+input_dir_all = []
```
修正の必要はないですが、このあたり、HashやRuby3.2からはいったDataクラスなどを使って

```ruby
[
  { input_string: 'xxxx', filenames: [ 'aaa.txt', 'bbb.txt'] },
  { input_string: 'yyyy', filenames: [ 'ccc.txt', 'ddd.txt'] },
]
```
のようにあらわしてあげるようにすると、indexアクセスによる煩雑さが少しは軽減されるかもしれません。

a-kuroki-gs replied:
ありがとうございます！
まさに自分でもどうしよう・・・ディレクトリとファイルで入力順とずれてしまうなあ・・・
と感じていたところでしたので、勉強になりました！

ただ、変更も大きくなるため、ひとまずこのまま進めてみようと思います！

yoshitsugu reviewed (approved):
OKです！一点コメントしましたので参考までに。
